src/combinatorics/colex.lean

```diff
@@ -0,0 +1,236 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+
+import data.finset
+import data.fintype.basic
+import algebra.geom_sum
+
+/-!
+# Colex
+
+We define the colex ordering for finite sets, and give a couple of important
+lemmas and properties relating to it.
+
+The colex ordering likes to avoid large values - it can be thought of on
+`finset ℕ` as the "binary" ordering. That is, order A based on
+`∑_{i ∈ A} 2^i`.
+It's defined here a slightly more general way, requiring only `has_lt α` in
+the definition of colex on `finset α`. In the context of the Kruskal-Katona
+theorem, we are interested in particular on how colex behaves for sets of a
+fixed size. If the size is 3, colex on ℕ starts
+123, 124, 134, 234, 125, 135, 235, 145, 245, 345, ...
+
+## Main statements
+* `colex_hom`: strictly monotone functions preserve colex
+* Colex order properties - linearity, decidability and so on.
+* `max_colex`: if A < B in colex, and everything in B is < t, then everything
+  in A is < t. This confirms the idea that an enumeration under colex will
+  exhaust all sets using elements < t before allowing t to be included.
+* `binary_iff`: colex for α = ℕ is the same as binary
+  (this also proves binary expansions are unique)
+
+## Notation
+We define `<ᶜ` and `≤ᶜ` to denote colex ordering, useful in particular when
+multiple orderings are available in context.
+
+## Tags
+colex, colexicographic, binary
+
+## References
+* http://b-mehta.github.io/maths-notes/iii/mich/combinatorics.pdf
+-/
+
+variable {α : Type*}
+
+open finset
+
+/--
+A <ᶜ B if the largest thing that's not in both sets is in B.
+In other words, max (A ▵ B) ∈ B.
+-/
+def colex_lt [has_lt α] (A B : finset α) : Prop :=
+  ∃ (k : α), (∀ {x}, k < x → (x ∈ A ↔ x ∈ B)) ∧ k ∉ A ∧ k ∈ B
```

Comment:
Usually we define `def finset.colex (α) := finset α` then add instances `[has_lt (finset.colex α)]` etc. This way you can use the usual `<`, `≤` etc.
You can also define `def finset.to_colex {α} (s : finset α) : finset.colex α := s` and prove that it is monotone etc.

Reply:
My concern with this method was that it'd be hard to use in practice, because when I have a set of finsets I sometimes care about both orderings, which is why I deliberately used an alternate `<` definition so that they can be distinguished

Reply:
Plus in the future I'd like to have other proofs which work by defining orderings and compressions like this one, where we'd define even more orderings on `finset α` and want to work with them together, so it'd be inconvenient to have this `finset.colex` wrapper

Reply:
What's wrong with `s.to_colex < t.to_colex`? You can have the same notation for this relation.

Reply:
Ooh didn't know about that! I agree that's a better solution.